10401b: simplification of Relationals

smichr opened the pull request:
ugh. The code is not simple. This is much worse than regular simplification if you try to keep a potentially invalid relationship from becoming valid, e.g. `Lt(x, x + 1)` is invalid if `x == I` so we cannot say that this relationship is automatically true. 

Some key features to keep are
1. to not reorganize the left and rhs (else an Eq that had something isolated on the lhs may no longer be that way after simplification, e.g. Eq(y, f(x)) could become Eq(y - f(x), 0))
2. to not return an expression whose character will differ from the original (e.g. not have the same singularities, etc...).

closes #10401 
closes #10404 
- [x] track down why the unrad solution changed

**References**

---
1. https://groups.google.com/forum/#!searchin/sympy/singularities$20limit/sympy/oaoAIp4DOxQ/cGZoWbf-ieUJ
2. https://github.com/sympy/sympy/wiki/Infinities-and-Singularities
3. https://github.com/sympy/sympy/issues/6091
4. https://github.com/sympy/sympy/pull/971

smichr committed:
more simplification of Relationals

Allow more simplification while retaining original character.

It's ok to simplify Lt(x, 2*x + 1) to x > -1 since the x does
not cancel. It is not ok to let Lt(x, x + 1) become True, however,
since x could be imaginary and then this relational would raise an
error.

smichr commented:
I think this can be made simpler. 

Some key features to keep are 

1. to not reorganize the left and rhs (else an `Eq` that had something isolated on the lhs may no longer be that way after simplification, e.g. `Eq(y, f(x))` could become `Eq(y - f(x), 0)`)
1. to not return an expression whose character will differ from the original (e.g. not have the same singularities, etc...).

smichr commented:
OK, I think this is done now.

smichr commented:
OK, I'm stuck. I don't know what the test passes in all but 2.7. Is there something wrong with the True, S.true test? Does _sympify(True) not return S.true in 2.7?

jksuom commented:
> Does _sympify(True) not return S.true in 2.7?

It does in my system.
```
In [8]: srepr(_sympify(True))
Out[8]: 'S.true'
```
Must be something else...

jksuom commented:
There was no problem:
```
sympy pr-10501$ bin/test test_relational
============================= test process starts ==============================
executable:         /usr/bin/python  (2.7.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 1.14
random seed:        69899449
hash randomization: on (PYTHONHASHSEED=1271816623)

sympy/core/tests/test_relational.py[37] .....fff............................. 
                                                                            [OK]

======== tests finished: 34 passed, 3 expected to fail, in 5.56 seconds ========
```
I guess you should just restart the job.

smichr commented:
Thanks for checking that out. I restarted. Let's see.

smichr commented:
Thanks to @pbrady and the `rerun` option for tests, I was able to quickly identify that it was a Float problem that was causing an equality to sometimes fail. I think I have that fixed now by converting Float to Rational.

jksuom commented on sympy/core/relational.py:
```diff
@@ -171,24 +171,171 @@ def equals(self, other, failing_expression=False):
+        # simplify as long as it does not remmove any denominators
+        denL = [d for d in denoms(L) if not d.is_number]
+        _L = L.simplify(ratio=ratio, measure=measure)
+        _denL = [d for d in denoms(L) if not d.is_number]
```
Perhaps this should contain ``denoms(_L)``.

smichr replied:
got it (and on _denR, too)

smichr commented:
This failing because after simplification, `factorial(x)/gamma(x)` is 1 and that looks like a denominator was lost. See note in test_simplify that was added with the modified test.

smichr committed:
allow new denominators to be introduced

This is failing because after simplification, `factorial(x)/gamma(x)` is x
and that looks like a denominator was lost. See note in test_simplify that
was added with the modified test.

smichr commented:
@jksuom, do you have an idea about how to handle this latest issue? I am trying to do "safe simplification" by not allowing simplify to return an expression which has lost a singularity defined by some denominator, e.g. `1/(1/x+1)` is not allowed to simplify to `x/(1 + x)`. But then this disallows `factorial(x)/gamma(x)` from simplifying to 1. I wonder if I should only disallow denominators which are **Add** instances to not disappear.

@jksuom wrote:

> 1/(1/x+1) is not allowed to simplify to x/(1 + x).

I'm afraid that trying to achieve this could result in problems hard to handle. In fact, there are reasons why I would rather allow this simplification.

BTW, you probably meant that ``factorial(x)/gamma(x)`` should simplify to ``x`` (for ``x`` a positive integer).

smichr commented:
> you probably meant that factorial(x)/gamma(x) should simplify to x

yes (the test was `Eq(trig identity, factorial(x)/gamma(x))` which simplifies to `Eq(x, 1)`).

smichr commented:
> there are reasons why I would rather allow this [`1/(1/x + 1)` -> `x/(x + 1)`] simplification

I would be interested to know what they are. The reason I am thinking to disallow this is that a Relational must be more careful than an expression that is being solved. In the latter case, one can use simplifications to find a solution and then *check* the solutions to eliminate any spurious roots. In the example I gave, the root of `x=0` would be rejected because it is not in the domain of the original function (please forgive mathematical inaccuracies, here). 

I was taking my cues from the you in this regard, btw.

If we do allow such simplifications then if we have `a = (1/x - x)*x` and `b=1 = x**2` (which is just `a.simplify()`) then `Eq(a - b, 0).simplify()` would be True -- and I see that it is in the PR(!) -- but for x=0 the lhs gives nan and the equality is False. And it is that which I am trying to avoid: giving a simplified result that would give a different answer than the non-simplified result.

Alternatively, we could return an **And** instance where condition under which the simplification is True are enumerated, e.g. using the `a` and `b` from the previous paragraph, `Eq(a - b, 0).simplify() -> And(True, Ne(x, 0)) -> Ne(x, 0)`.

jksuom commented:
This is something that I haven't figured out in detail. It's more like a feeling, but I'll try to collect my thoughts.

``a = (1/x - x)*x`` and ``b = 1 - x**2`` are two different expressions, but I think they should be considered as denoting the same (rational) function. It is true that ``a`` gives ``nan`` at ``x = 0`` but that could be interpreted as "I was not able to compute the value", not as "there is no appropriate value". I think the evaluation of ``a`` should be improved to recognize 0 as a removable singularity and return the correct limit value 1. (This would be analogous to the way e.g. boolean expressions and integrals are often treated: if one method returns None, then try another one.)

In brief: I think that it would be preferable to improve the evaluation of expressions. There appear to be no clear grounds for making distinction between expressions denoting essentially same (abstract) objects.

smichr commented:
@jksuom, are you expressing a different feeling now relative to the one that you expressed originally in your comment at #10401 ?  (Sorry, I don't know how to grab the link to send you directly there, but there is only one comment on the issue :-))

jksuom commented:
That was about ``x/(x + 1)``. In fact, I was also then mentally identifying it with ``1/(1 + 1/x)``, and since the latter takes the value 1, I was feeling uneasy about the former. It is true that the value 1 is attained only at ``zoo``, but that seems to be generally not excluded (unless ``x = Symbol('x', finite=True)``).

smichr commented:
But you said "of course, f != g" (where `f = x/(x + 1)` and `g = 1/(1 + 1/x)`. That they aren't the same is why I think we can't let the `Eq(f, g)` simplify to True; it is also why we don't report 0 as a solution to the one but do for the other:

```python
>>> var('x');g=1/(1/x+1)
x
>>> solve(g)
[]
>>> solve(g.simplify())
[0]
```

jksuom commented:
> of course, f != g

I understood it was saying that f and g are *structurally* different. Anyway, that was the result when I tested it :-)

It is true that simplifying ``Eq(f, g)`` to True would currently lead to conflicts. But I was hoping that ``solve(g)`` could eventually be modified to yield ``[0]`` as ``g`` denotes a rational function.

smichr commented:
@asmeurer , I remember some time ago discussing whether 0 is a solution to something like `1/(1/x + 1)` but have only found the references that I list in the OP. Do you recall anything else about the discussion or where it took place? I thought it had to do with whether or not a removable singularity was to be considered a solution to an equation.

smichr commented:
OK, allowing removable singularities to disappear will simplify things, I think. In addition, I think I have a way to handle the case where a ratio of functions (like factorial and gamma) need to simplify. I'll try it some time this week.
